exec/cnex/lib/sqlite.c

```diff
@@ -0,0 +1,404 @@
+            goto cleanup;
+        }
+        r = sqlite3_bind_text(stmt, c, TCSTR(parameters->dictionary->data[i].value->string), -1, SQLITE_TRANSIENT);
+        if (r != SQLITE_OK) {
+            exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+            goto cleanup;
+        }
+    }
+    int columns = sqlite3_column_count(stmt);
+    for (;;) {
+        r = sqlite3_step(stmt);
+        if (r == SQLITE_DONE) {
+            break;
+        }
+        if (r == SQLITE_ROW) {
+            Cell *row = cell_createArrayCell(0);
+            for (int i = 0; i < columns; i++) {
+                cell_arrayAppendElementPointer(row, cell_fromCString((const char *)sqlite3_column_text(stmt, i)));
+            }
+            cell_arrayAppendElementPointer(rows, row);
+        } else {
+            exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+            goto cleanup;
+        }
+    }
+    push(exec->stack, rows);
+
+cleanup:
+    sqlite3_finalize(stmt);
+    free(sql);
+    cell_freeCell(parameters);
+}
+
+void sqlite_execOne(TExecutor *exec)
+{
+    Cell *parameters = cell_fromCell(top(exec->stack)); pop(exec->stack);
+    char *sql = string_asCString(top(exec->stack)->string); pop(exec->stack);
+    Object *db = check_database(top(exec->stack)); pop(exec->stack);
+
+    if (db == NULL) {
+        exec->rtl_raise(exec, "SqliteException.InvalidDatabase", "", BID_ZERO);
+        cell_freeCell(parameters);
+        free(sql);
+        return;
+    }
+
+    Cell *result = cell_createArrayCell(0);
+    sqlite3_stmt *stmt = NULL;
+    BOOL Retval = FALSE;
+
+    int r = sqlite3_prepare_v2(db->ptr, sql, -1, &stmt, NULL);
+    if (r != SQLITE_OK) {
+        exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+        goto done;
+    }
+    for (int i = 0; i < parameters->dictionary->len; i++) {
+        int c = sqlite3_bind_parameter_index(stmt, TCSTR(parameters->dictionary->data[i].key));
+        if (c == 0) {
+            exec->rtl_raise(exec, "SqlException.ParameterName", TCSTR(parameters->dictionary->data[i].key), BID_ZERO);
+            goto done;
+        }
+        r = sqlite3_bind_text(stmt, c, TCSTR(parameters->dictionary->data[i].value->string), -1, SQLITE_TRANSIENT);
+        if (r != SQLITE_OK) {
+            exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+            goto done;
+        }
+    }
+    int columns = sqlite3_column_count(stmt);
+    r = sqlite3_step(stmt);
+    if (r == SQLITE_DONE) {
+        goto done;
+    }
+    if (r == SQLITE_ROW) {
+        for (int i = 0; i < columns; i++) {
+            cell_arrayAppendElementPointer(result, cell_fromCString((const char *)sqlite3_column_text(stmt, i)));
+        }
+    } else {
+        exec->rtl_raise(exec, "SqlException", sqlite3_errmsg(db->ptr), number_from_sint32(r));
+        goto done;
+    }
+    Retval = TRUE;
+
+done:
```

Comment:
There's considerable inconsistency in the naming of these goto labels. The previous function uses `cleanup`, this one uses `done`, and the next function doesn't use one at all (but it could). Suggest making these consistent.

Reply:
I tried to only use them when necessary, and name them for what they actually were in there for.  Would it be bad to use a "bail" tag to cover all cases?